crates/ra_fmt2/src/edit_tree.rs

```diff
@@ -0,0 +1,115 @@
+use crate::dsl::{SpacingDsl, SpacingRule, SpaceLoc, SpaceValue};
+use crate::pattern::{Pattern, PatternSet};
+use crate::rules::spacing;
+use crate::trav_util::{walk, walk_nodes, walk_tokens};
+
+use ra_syntax::{
+    NodeOrToken, SmolStr, SyntaxElement,
+    SyntaxKind::{self, *},
+    SyntaxNode, SyntaxToken, TextRange, TextUnit, WalkEvent, T,
+};
+
+use std::collections::{HashMap, HashSet};
+
+// TODO make more like intellij's fmt model
+// Model holds immutable tree and mutable intermediate model to produce diff
+// the model will probably have to create its own tree to add the extra
+// info to each token/node:
+//
+// [1,2,3];
+// can be Brace token, ident, comma all of which knows their own rules and apply
+// them accordingly to produce [1, 2, 3]; ???
+
+#[derive(Debug)]
+// this will probably be an enum that holds a struct for space/indent/ect
+struct AtomEdit {
+    space_loc: SpaceLoc,
+    val: SpaceValue,
+    start: TextUnit,
+}
+
+impl AtomEdit {}
+
+#[derive(Debug)]
+pub(crate) struct SynBlock {
+    //indent: some enum?
+    kind: SyntaxKind,
+    node: SyntaxNode,
+    tokens: Vec<SyntaxToken>,
+    range: TextRange,
+    edits: Vec<AtomEdit>,
+    ws_rules: Vec<SpacingRule>,
```

Comment:
For spacing, I think we should follow IntelliJ's model and have a `spacing(child1, child2)` method:

```java
  /**
   * Returns a spacing object indicating what spaces and/or line breaks are added between two
   * specified children of this block.
   *
   * @param child1 the first child for which spacing is requested;
   *               {@code null} if given {@code 'child2'} block is the first document block
   * @param child2 the second child for which spacing is requested.
   * @return the spacing instance, or null if no special spacing is required. If null is returned,
   *         the formatter does not insert or delete spaces between the child blocks, but may insert
   *         a line break if the line wraps at the position between the child blocks.
   * @see Spacing#createSpacing(int, int, int, boolean, int)
   * @see Spacing#getReadOnlySpacing()
   */
  @Nullable
  Spacing getSpacing(@Nullable Block child1, @NotNull Block child2);
```